lib/parsers/transform.js

```diff
@@ -0,0 +1,519 @@
+  for (const tpl of templateInfos) {
+    const range = tpl.range;
+    const template = code.slice(...range);
+    const docLines = new DocumentLines(template);
+    const ast = glimmer.preprocess(template, { mode: 'codemod' });
+    ast.tokens = tokenize(code.slice(...tpl.templateRange), codeLines, tpl.templateRange[0]);
+    const allNodes = [];
+    glimmer.traverse(ast, {
+      All(node, path) {
+        const n = node;
+        n.parent = path.parentNode;
+        allNodes.push(node);
+        if (node.type === 'CommentStatement' || node.type === 'MustacheCommentStatement') {
+          comments.push(node);
+        }
+        if (node.type === 'TextNode') {
+          n.value = node.chars;
+          textNodes.push(node);
+        }
+      },
+    });
+    ast.content = template;
+    const allNodeTypes = new Set();
+    for (const n of allNodes) {
+      if (n.type === 'PathExpression') {
+        n.head.range = [
+          range[0] + docLines.positionToOffset(n.head.loc.start),
+          range[0] + docLines.positionToOffset(n.head.loc.end),
+        ];
+        n.head.loc = {
+          start: codeLines.offsetToPosition(n.head.range[0]),
+          end: codeLines.offsetToPosition(n.head.range[1]),
+        };
+      }
+      n.range =
+        n.type === 'Template'
+          ? [tpl.templateRange[0], tpl.templateRange[1]]
+          : [
+              range[0] + docLines.positionToOffset(n.loc.start),
+              range[0] + docLines.positionToOffset(n.loc.end),
+            ];
+
+      n.start = n.range[0];
+      n.end = n.range[1];
+      n.loc = {
+        start: codeLines.offsetToPosition(n.range[0]),
+        end: codeLines.offsetToPosition(n.range[1]),
+      };
+      if (n.type === 'Template') {
+        n.loc.start = codeLines.offsetToPosition(tpl.templateRange[0]);
+        n.loc.end = codeLines.offsetToPosition(tpl.templateRange[1]);
+      }
+      // split up element node into sub nodes to be able to reference tag name
+      // parts <Foo.Bar /> -> nodes for `Foo` and `Bar`
+      if (n.type === 'ElementNode') {
+        n.name = n.tag;
+        n.parts = [];
+        let start = n.range[0];
+        let codeSlice = code.slice(...n.range);
+        for (const part of n.tag.split('.')) {
+          const regex = new RegExp(`\\b${part}\\b`);
+          const match = codeSlice.match(regex);
+          const range = [start + match.index, 0];
+          range[1] = range[0] + part.length;
+          codeSlice = code.slice(range[1], n.range[1]);
+          start = range[1];
+          n.parts.push({
+            type: 'GlimmerElementNodePart',
+            name: part,
+            range,
+            parent: n,
+            loc: {
+              start: codeLines.offsetToPosition(range[0]),
+              end: codeLines.offsetToPosition(range[1]),
+            },
+          });
+        }
+      }
+      // block params do not have location information
+      // add our own nodes so we can reference them
```

Comment:
why do we do anything with block params? the whole glimmer AST should not be needed for gjs/gts parsing

Reply:
I reorganized the files as parser file was getting to big

Reply:
but that doesn't answer the question :sweat_smile:

Reply:
Ah, it's to have the rules no-undef and no-unused work in gjs/gts files. Like this we have the exact location of no-undef. 
And it also works for block params. Marking them unused when not used

Reply:
You can see it in thte tests for the parser

Reply:
we only have this configuration for gjs/gts in template-lint: https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/config/recommended.js#L98

Reply:
> we only have this configuration for gjs/gts in template-lint: https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/config/recommended.js#L98

Right, template-lint doesn't handle js scope. Here it can handle both, so the rules of template-lint could theoretically be moved to eslint. 

It also is more code to supporting other rules that are token based

Reply:
https://github.com/glimmerjs/glimmer-vm/issues/1490

Reply:
I will happily take PRs on that repo that improve the parser!